File: derma-clinic-twacha/server/lib/activity.js
/* =============================================================================
   Audit trail. Every write that a person could later dispute gets a row here:
   who, what, when, from where. Money and patient records especially.
   ========================================================================== */

import { run, all } from '../db.js';
import { clientIp } from './http.js';

export function log(req, action, entity = null, entityId = null, detail = null) {
  try {
    run(
      `INSERT INTO activity_log (user_id, action, entity, entity_id, detail, ip)
       VALUES (?, ?, ?, ?, ?, ?)`,
      req.user?.id ?? null, action, entity, entityId,
      detail == null ? null : JSON.stringify(detail), clientIp(req)
    );
  } catch {
    /* An audit failure must never take down the operation being audited. */
  }
}

export const recent = (limit = 100) => all(
  `SELECT a.*, u.name AS user_name
     FROM activity_log a LEFT JOIN users u ON u.id = a.user_id
    ORDER BY a.created_at DESC, a.id DESC
    LIMIT ?`,
  Math.min(Number(limit) || 100, 500)
);

File: studio-aarohi/core/activity.py
"""Audit trail — so the owner can see what changed and when."""

from .db import insert, now, q


def log(action: str, entity: str = "", entity_id=None, summary: str = "", user=None) -> None:
    name = "system"
    if user is not None:
        try:
            name = user["name"] or user["email"]
        except (TypeError, KeyError):
            name = str(user)
    insert("activity", {
        "user_name": name,
        "action": action,
        "entity": entity,
        "entity_id": entity_id,
        "summary": summary[:400],
        "created_at": now(),
    })


def recent(limit: int = 12) -> list:
    return q("SELECT * FROM activity ORDER BY id DESC LIMIT ?", (limit,))


def trim(keep: int = 500) -> None:
    q("DELETE FROM activity WHERE id NOT IN "
      "(SELECT id FROM activity ORDER BY id DESC LIMIT ?)", (keep,))
